File: orkbasecxx/serializers/SingleLineSerializer.cpp
/*
 * Oreka -- A media capture and retrieval platform
 * 
 * Copyright (C) 2005, orecx LLC
 *
 * http://www.orecx.com
 *
 * This program is free software, distributed under the terms of
 * the GNU General Public License.
 * Please refer to http://www.gnu.org/copyleft/gpl.html
 *
 */

#include "ace/OS_NS_ctype.h"
#include "Utils.h"
#include "SingleLineSerializer.h"


void SingleLineSerializer::AddString(const char* key, CStdString& value)
{
	CStdString pair;
	CStdString escapedValue;
	EscapeSingleLine(value, escapedValue);
	pair.Format("%s=%s ", key, (PCSTR)escapedValue);
	m_output += pair;
}

CStdString SingleLineSerializer::Serialize()
{
	m_deSerialize = false;		// Set Serialize mode
	m_object->Define(this);
	return m_output;
}

typedef enum {SingleLineStartState, SingleLineKeyState, SingleLineValueState, SingleLineErrorState} SingleLineState;

void SingleLineSerializer::DeSerialize(CStdString& input)
{
	// read string and extract values into map
	SingleLineState state = SingleLineStartState;
	CStdString key;
	CStdString value;
	// #### Additionally, errorDescription should be converted to Exceptions
	CStdString errorDescription;

	input.Trim();

	for(unsigned int i=0; i<input.length() && state!= SingleLineErrorState; i++)
	{
		TCHAR character = input[i];

		switch(state)
		{
		case SingleLineStartState:
			if(character == ' ')
			{
				;	// ignore spaces
			}
			else if ( ACE_OS::ace_isalnum(character) )
			{
				state = SingleLineKeyState;
				key = character;
			}
			else
			{
				state = SingleLineErrorState;
				errorDescription = "Cannot find key start, keys must be alphanum";
			}
			break;
		case SingleLineKeyState:
			if( ACE_OS::ace_isalnum(character) )
			{
				key += character;
			}
			else if (character == '=')
			{
				state = SingleLineValueState;
				value.Empty();
			}
			else
			{
				state = SingleLineErrorState;
				errorDescription = "Invalid key character, keys must be alphanum";
			}
			break;
		case SingleLineValueState:
			if( character == '=')
			{
				state = SingleLineErrorState;
				errorDescription = "Value followed by = sign, value should always be followed by space sign";				
			}
			else if (character == ' ')
			{
				state = SingleLineStartState;
			}
			else
			{
				value += character;
			}
			break;
		default:
			state = SingleLineErrorState;
			errorDescription = "Non-existing state";
		}	// switch(state)

		if ( (state == SingleLineStartState) || (i == (input.length()-1)) )
		{
			if (!key.IsEmpty())
			{
				// SingleLine unescape
				CStdString unescapedValue;
				UnEscapeSingleLine(value, unescapedValue);

				// Add pair to key-value map
				m_map.insert(std::make_pair(key, unescapedValue));
				key.Empty();
				value.Empty();
			}
		}
	}	// for(int i=0; i<input.length() && state!= SingleLineErrorState; i++)

	Serializer::DeSerialize();
}

// Escape the space, equals and percent characters for serializing to Key-Value-Pair text
void SingleLineSerializer::EscapeSingleLine(CStdString& in, CStdString& out)
{
	for(unsigned int i=0; i<in.length();i++)
	{
		TCHAR c = in[i];
		if (c == ' ')
		{
			out+= "%s";
		}
		else if (c == '%')
		{
			out+= "%p";
		}
		else if (c == '=')
		{
			out+= "%e";
		}
		else
		{
			out+= c;
		}
	}
}

// Unescape the space, equals and percent characters for serializing to Key-Value-Pair text
void SingleLineSerializer::UnEscapeSingleLine(CStdString& in, CStdString& out)
{
	unsigned int iin = 0;

	while(iin<in.length())
	{ 
		if ( in[iin] == '%')
		{
			iin++;

			switch (in[iin])
			{
			case 's':
				out += ' ';
				break;
			case 'p':
				out += '%';
				break;
			case 'e':
				out += '=';
				break;
			}
		}
		else
		{
			out += in[iin];
		}
		iin++;
	}
}

CStdString SingleLineSerializer::FindClass(CStdString& input)
{
	CStdString result;
	int equalsPostion = input.Find('=');
	if (equalsPostion != -1)
	{
		int spacePostion = input.Find(' ');
		if (spacePostion > equalsPostion)
		{
			result = input.Mid(equalsPostion+1, spacePostion-equalsPostion-1);
		}
		else
		{
			result = input.Mid(equalsPostion+1, input.GetLength() - equalsPostion - 1);
		}
	}
	result.ToLower();
	return result;
}

File: orkbasej/src/main/java/net/sf/oreka/serializers/SingleLineSerializer.java
/*
 * Oreka -- A media capture and retrieval platform
 * 
 * Copyright (C) 2005, orecx LLC
 *
 * http://www.orecx.com
 *
 * This program is free software, distributed under the terms of
 * the GNU General Public License.
 * Please refer to http://www.gnu.org/copyleft/gpl.html
 *
 */

package net.sf.oreka.serializers;

import net.sf.oreka.OrkException;
import net.sf.oreka.OrkObject;
import net.sf.oreka.OrkObjectFactory;

/** 
 * Serializer that generates and parses objects to and from a single line of text.
 * key-value pairs are separated by spaces.
 * key and values are separated by equal signs.
 * example: message=doit what=run 
 */
public class SingleLineSerializer extends KeyValueSerializer
{
	StringBuffer output;
	
	public String serialize(OrkObject input) throws OrkException
	{
		output = new StringBuffer();
		deserialize = false;		// Set Serialize mode
		addClassName(input.getOrkClassName());
		input.define(this);
		return output.toString();
	}
	
	enum SingleLineState {StartState, KeyState, ValueState};
	
	public OrkObject deSerialize(String input) throws OrkException
	{
		// read string and extract values into map
		input = input.trim();

		SingleLineState state = SingleLineState.StartState;

		StringBuffer key = new StringBuffer();
		StringBuffer value = new StringBuffer();
		String firstValue = null;
		
		for(int i=0; i<input.length(); i++)
		{
			Character character = input.charAt(i);

			switch(state)
			{
			case StartState:
				if(character == ' ')
				{
					;	// ignore spaces
				}
				
				else if ( Character.isLetterOrDigit(character) )
				{
					state = SingleLineState.KeyState;
					key.append(character);
				}
				else
				{
					throw new OrkException("SingleLineSerializer.deSerialize: Cannot find key start, keys must be alphanum, wrong char:"  + character + " input:" + input);
				}
				break;
				
			case KeyState:
				if( Character.isLetterOrDigit(character) )
				{
					key.append(character);
				}
				else if (character == '=')
				{
					state = SingleLineState.ValueState;
					value = new StringBuffer();
				}
				else
				{
					throw new OrkException("SingleLineSerializer.deSerialize: Invalid key character, keys must be alphanum, wrong char:"  + character + " input:" + input);
				}
				break;
			case ValueState:
				if( character == '=')
				{
					throw new OrkException("SingleLineSerializer.deSerialize: equals sign misplaced at index:" + i + " in input:" + input);				
				}
				else if (character == ' ')
				{
					state = SingleLineState.StartState;
				}
				else
				{
					value.append(character);
				}
				break;
			}	// switch(state)

			if ( (state == SingleLineState.StartState) || (i == (input.length()-1)) )
			{
				if (key.length() > 0)
				{
					// Add pair to key-value map
					map.put(key.toString().toLowerCase(), unEscapeSingleLine(value.toString()));
					if (firstValue == null)
					{
						firstValue = value.toString().toLowerCase();
					}
					key = new StringBuffer();
					value = new StringBuffer();
				}
			}
		}

		deserialize = true;		// Set DeSerialize mode
		
		OrkObject obj = OrkObjectFactory.instance().newOrkObject(firstValue);
		obj.define(this);
		obj.validate();
		return obj;
	}
	
	public void addClassName(String value)
	{
		addString(classKey, value);
	}
	
	void addString(String key, String value)
	{
		String pair;
		String escapedValue = escapeSingleLine(value);
		pair = key + "=" + escapedValue + " ";
		output.append(pair);
	}

	/**
	 * Finds the Object Class from a serialized string. 
	 * i.e. for a serialized object of the form:
	 * 		message=dosomething when=now
	 * it returns "dosomething"
	 * 
	 * @param input
	 * @return
	 */
	/*
	static String findClass(String input)
	{
		String result = "";
		int equalsPosition = input.indexOf("=");
		if (equalsPosition != -1)
		{
			int spacePostion = input.indexOf(" ");
			if (spacePostion > equalsPosition)
			{
				result = input.substring(equalsPosition+1, spacePostion-1);
			}
			else
			{
				result = input.substring(equalsPosition+1, input.length() - 1);
			}
		}
		return result.toLowerCase();
	}
	*/
	
	String escapeSingleLine(String in)
	{		
		StringBuffer out = new StringBuffer();
		for(int i=0; i<in.length();i++)
		{
			char c = in.charAt(i);
			if (c == ' ')
			{
				out.append("%s");
			}
			else if (c == ':')
			{
				out.append("%c");
			}
			else if (c == '%')
			{
				out.append("%p");
			}
			else
			{
				out.append(c);
			}
		}		
		return out.toString();
	}
	
	String unEscapeSingleLine(String in)
	{
		StringBuffer out = new StringBuffer();
		int iin = 0;

		while(iin<in.length())
		{ 
			if ( in.charAt(iin) == '%')
			{
				iin++;

				switch (in.charAt(iin))
				{
				case 's':
					out.append(' ');
					break;
				case 'c':
					out.append(':');
					break;
				case 'p':
					out.append('%');
					break;
				}
			}
			else
			{
				out.append(in.charAt(iin));
			}
			iin++;
		}
		return out.toString();
	}
}
